homeassistant/components/tag/__init__.py

```diff
@@ -133,7 +189,75 @@ async def async_scan_tag(
     )
 
     if tag_id in helper.data:
-        await helper.async_update_item(tag_id, {LAST_SCANNED: dt_util.utcnow()})
+        await helper.async_update_item(
+            tag_id, {LAST_SCANNED: dt_util.utcnow(), DEVICE_ID: device_id or ""}
+        )
     else:
-        await helper.async_create_item({TAG_ID: tag_id, LAST_SCANNED: dt_util.utcnow()})
+        await helper.async_create_item(
+            {TAG_ID: tag_id, LAST_SCANNED: dt_util.utcnow(), DEVICE_ID: device_id or ""}
+        )
     _LOGGER.debug("Tag: %s scanned by device: %s", tag_id, device_id)
+
+
+class TagEntity(Entity):
+    """Representation of a Tag entity."""
+
+    _unrecorded_attributes = frozenset({TAG_ID})
+    _attr_translation_key = DOMAIN
+    _attr_should_poll = False
+
+    # Implements it's own platform
+    _no_platform_reported = True
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        name: str,
+        tag_id: str,
+        last_scanned: str,
+        device_id: str | None,
+    ) -> None:
+        """Initialize the Tag event."""
+        self.entity_id = f"tag.{slugify(name)}"
+        self.hass = hass
+        self._attr_name = name
+        self._tag_id = tag_id
+        self._last_device_id: str | None = device_id
+        self._last_scanned = last_scanned
+        self._attr_unique_id = tag_id
+
+        self._state_info = {
+            "unrecorded_attributes": self._Entity__combined_unrecorded_attributes  # type: ignore[attr-defined]
+        }
+
+    @callback
+    def async_handle_event(self, device_id: str | None, last_scanned: str) -> None:
+        """Handle the Tag scan event."""
+        if _LOGGER.isEnabledFor(logging.DEBUG):
+            _LOGGER.debug(
+                "Tag scanned for %s with device %s last scanned before %s and scanned now %s",
+                self._tag_id,
+                device_id,
+                self._last_scanned,
+                last_scanned,
+            )
+        self._last_device_id = device_id
+        self._last_scanned = last_scanned
+        self.async_write_ha_state()
+
+    @property
+    @final
+    def state(self) -> str | None:
+        """Return the entity state."""
+        if (last_scanned := dt_util.parse_datetime(self._last_scanned)) is None:
+            return None
+        return last_scanned.isoformat(timespec="milliseconds")
+
+    @property
+    def extra_state_attributes(self) -> dict[str, Any]:
+        """Return the state attributes of the sun."""
+        return {TAG_ID: self._tag_id, DEVICE_ID: self._last_device_id}
```

Comment:
How can the user make use of the device id attribute?
Also, maybe instead of calling it `device_id`, I think it should be clear from the attribute key it's the ID of the device which scanned the tag.

Reply:
> How can the user make use of the device id attribute?

As also noted in the tag documentation you may want in an automation to have different actions depending on the device that triggered the tag.

> Also, maybe instead of calling it `device_id`, I think it should be clear from the attribute key it's the ID of the device which scanned the tag.

Not sure what it should be called then? `device_id` is already in the documentation with an explanation https://www.home-assistant.io/integrations/tag/#tag-scanned-events

Reply:
Yeah, but with a `device_id` key inside a `tag_scanned` event, there's context which is lost when the device id is a state attribute.
I think the attribute should have `scanned` in its name, for example `last_scanned_device_id` or `last_scanned_by_device_id`.

Reply:
Makes sense 👍